internal/ext/unsafex/unsafex.go

```diff
@@ -17,42 +17,124 @@
 // Importing this package should be treated as equivalent to importing unsafe.
 package unsafex
 
-import "unsafe"
-
-// Size is like [unsafe.Sizeof], but it is a generic function and it returns
-// an int instead of a uintptr (Go does not have types so large they would
-// overflow an int).
-func Size[T any]() int {
-	var zero T
-	return int(unsafe.Sizeof(zero))
+import (
+	"fmt"
+	"unsafe"
+)
+
+// Int is a constraint for any integer type.
+type Int interface {
+	~int8 | ~int16 | ~int32 | ~int64 | ~int |
+		~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uint |
+		~uintptr
+}
+
+// Layout is the layout of a type.
+//
+// This is a more convenient abstraction that manipulating the size and
+// alignment separately.
+type Layout struct {
+	Size, Align int
+
+	// TODO: Add something like PointerFree? It's not clear if Go even exposes
+	// an easy way to compute this. There's certainly no intrinsic for it...
 }
 
-// Index is like [unsafe.Add], but it operates on a typed pointer and scales the
+// LayoutOf returns the layout of some type.
+func LayoutOf[T any]() Layout {
+	var v T
+	return Layout{
+		Size:  int(unsafe.Sizeof(v)),
+		Align: int(unsafe.Alignof(v)),
+	}
+}
+
+// Add is like [unsafe.Add], but it operates on a typed pointer and scales the
 // offset by that type's size, similar to pointer arithmetic in Rust or C.
 //
 // This function has the same safety caveats as [unsafe.Add].
-func Index[T any](p *T, idx int) *T {
+//
+//go:nosplit
+func Add[P ~*E, E any, I Int](p P, idx I) P {
 	raw := unsafe.Pointer(p)
-	raw = unsafe.Add(raw, idx*Size[T]())
-	return (*T)(raw)
+	raw = unsafe.Add(raw, int(idx)*LayoutOf[E]().Size)
+	return P(raw)
 }
 
-// Cast performs an unchecked cast of one pointer type to another, through
-// an [unsafe.Pointer].
-func Cast[To, From any](p *From) *To {
-	return (*To)(unsafe.Pointer(p))
-}
-
-// Transmute bit-casts a value of type From to a value of type To.
+// Bitcast bit-casts a value of type From to a value of type To.
 //
 // This operation is very dangerous, because it can be used to break package
 // export barriers, read uninitialized memory, and forge pointers in violation
 // of [unsafe.Pointer]'s contract, resulting in memory errors in the GC.
 //
-// Transmute should only be used on types the caller owns, and those types
-// should have identical GC shapes.
-func Transmute[To, From any](p From) To {
-	return *Cast[To](&p)
+// Panics if To and From have different sizes.
+//
+//go:nosplit
+func Bitcast[To, From any](v From) To {
```

Comment:
This feels significantly more complicated than the previous implementation of `Transmute`. Is that really necessary? `Transmute` was far more clear to read. Did it have material safety or perf issues? If so, please elaborate (maybe even in comments, to explain why the method body isn't the simpler, more intuitive implementation).